// @goodt/eslint-config-base/widgets/index.js
module.exports = {
    extends: ['@goodt/eslint-config-base'],
    rules: {
        // eslint common
        'no-console': ['error', { allow: ['warn', 'error'] }],
        'no-param-reassign': ['error', { props: false }],
        curly: ['error', 'all'],
        'no-magic-numbers': [
            'warn',
            { ignoreArrayIndexes: true, detectObjects: true, ignore: [0, 1, -1, 100, 1000, 60] }
        ],
        'no-unused-expressions': ['warn'],
        'id-length': [
            'error',
            {
                min: 3,
                exceptions: [
                    'a',
                    'b',
                    'x',
                    'y',
                    'i',
                    'j',
                    'n',
                    'e',
                    '_',
                    '__',
                    '$c',
                    '$t',
                    'id',
                    'vm',
                    'me',
                    'ns',
                    'tr',
                    'td',
                    'li',
                    'ul',
                    'to',
                    'on',
                    'or',
                    'be',
                    'el',
                    'px'
                ]
            }
        ],
        eqeqeq: ['error', 'smart'],
        'prefer-destructuring': [
            'error',
            {
                VariableDeclarator: {
                    array: false,
                    object: true
                },
                AssignmentExpression: {
                    array: false,
                    object: true
                }
            },
            {
                enforceForRenamedProperties: false
            }
        ],
        'no-underscore-dangle': [
            'error',
            {
                allowAfterSuper: true,
                allowAfterThis: true,
                enforceInMethodNames: false,
                allowAfterThisConstructor: true
            }
        ],
        'class-methods-use-this': 'off',

        // vue
        'vue/no-empty-pattern': 'error',
        'vue/no-constant-condition': 'error',
        'vue/attributes-order': [
            'off',
            {
                order: [
                    'CONDITIONALS',
                    'LIST_RENDERING',
                    'GLOBAL',
                    'DEFINITION',
                    'OTHER_ATTR',
                    'RENDER_MODIFIERS',
                    ['UNIQUE', 'SLOT'],
                    'TWO_WAY_BINDING',
                    'OTHER_DIRECTIVES',
                    'EVENTS',
                    'CONTENT'
                ],
                alphabetical: false
            }
        ],
        'vue/order-in-components': [
            'error',
            {
                order: [
                    'el',
                    'name',
                    'key',
                    'parent',
                    'functional',
                    ['delimiters', 'comments'],
                    'extends',
                    ['components', 'directives', 'filters'],
                    'mixins',
                    ['provide', 'inject'],
                    'ROUTER_GUARDS',
                    'layout',
                    'middleware',
                    'validate',
                    'scrollToTop',
                    'transition',
                    'loading',
                    'inheritAttrs',
                    'model',
                    ['props', 'propsData'],
                    'emits',
                    'setup',
                    'asyncData',
                    'data',
                    'fetch',
                    'head',
                    'computed',
                    'watch',
                    'watchQuery',
                    'LIFECYCLE_HOOKS',
                    'methods',
                    ['template', 'render'],
                    'renderError'
                ]
            }
        ],
        'vue/component-tags-order': [
            'error',
            {
                order: ['template', ['style', 'script']]
            }
        ],
        'vue/component-name-in-template-casing': [
            'error',
            'kebab-case',
            {
                registeredComponentsOnly: true,
                ignores: []
            }
        ],
        'vue/max-attributes-per-line': [
            'error',
            {
                singleline: 5,
                multiline: 1
            }
        ],

        // unicorn
        'unicorn/prevent-abbreviations': 'off',
        'unicorn/explicit-length-check': 'error',
        'unicorn/prefer-ternary': 'error',
        'unicorn/prefer-array-some': 'error',
        'unicorn/prefer-array-flat': 'error',
        'unicorn/prefer-includes': 'error',

        // radar
        'radar/prefer-single-boolean-return': 'error',
        'radar/no-use-of-empty-return-value': 'error',
        'radar/no-all-duplicated-branches': 'error',
        'radar/no-identical-conditions': 'error',

        // goodt
        'goodt-rules/component-name-regex': 'error',

        // more
        'more/prefer-includes': 0
    }
};
