// @goodt/webpack/remote/config/module-federation.js
const { getShared } = require('@goodt-webpack/common/shared');
const AutomaticVendorFederation = require('@module-federation/automatic-vendor-federation');
const {
    container: { ModuleFederationPlugin }
} = require('webpack');

module.exports = ({ packageJson, packageLockJson, sharedConfig, exposes, entryFile, useShared, shareWcorePackages }) => {
    let shared = {
        ...AutomaticVendorFederation({
            // exclude is mandatory prop
            exclude: [ ...packageJson.sharedDependencies, '@goodt/common-deps'],
            packageJson,
            shareFrom: ['dependencies', 'peerDependencies']
        }),
        ...(useShared && getShared({ packageJson, packageLockJson, shareWcorePackages })),
        ...sharedConfig
    };

    return new ModuleFederationPlugin({
        library: { type: "module" },
        name: packageJson.name,
        filename: entryFile,
        exposes,
        // This should be our 'shared contract' between the host and the remotes.
        shared,
    });
};
