// @goodt/webpack/remote/config/optimization.js
const getWidgetName = (module) => {
    const regExp = /\b(Elem[A-z]+)\b/;
    const { context } = module;
    const strMatchElem = context.match(regExp);
    if (strMatchElem != null && strMatchElem.length > 1) {
        return strMatchElem[1];
    }
    return false;
};

const getWidgetPanelName = (module) => {
    const regExp = /\b(Elem[A-z]+)\b/;
    const strMatchElem = module.resource.match(regExp);
    if (strMatchElem != null && strMatchElem.length > 1) {
        return strMatchElem[1] + '~' + 'panels';
    }
    return false;
};

module.exports = ({ packageJson } = {}) => {
        // Требуется для сборок `ремоутов`, чтобы повторно не грузить
        // `goodt-wcore, vue` вместе с чанком своих специфических `vendors`
        // для `хоста` в целом не принципиально
        const npmCacheGroups = packageJson.sharedDependencies.reduce(
            (acc, name) => ({
                [name]: {
                    name: `shared.${name}`,
                    test: new RegExp(`/node_modules/${name}/`),
                    priority: 40
                    // reuseExistingChunk: true //?
                },
                ...acc
            }),
            {}
        );
        // специфика ремоутов (виджетов)
        const cacheGroups = {
            ...npmCacheGroups,
            widgets: {
                chunks: 'async',
                test: /\/src\/.*\b(Elem[A-z]+)\b/,
                // enforce: true,
                maxSize: 300 * 1000,
                minSize: 0,
                priority: 30,
                reuseExistingChunk: true,
                name(module) {
                    return getWidgetName(module);
                }
            },
            panels: {
                chunks: 'async',
                test: /\/src\/.*\b(Elem[A-z]+)\b.*\b[A-z]+Panel\.vue/,
                maxSize: 300 * 1000,
                minSize: 0,
                reuseExistingChunk: true,
                priority: 35,
                name(module) {
                    return getWidgetPanelName(module);
                }
            }
        };

    return { splitChunks: { cacheGroups } };
};
