// @goodt/webpack/remote/config/webpack.common.js
// LIBS
const path = require('path');
const { merge } = require('webpack-merge');
// PLUGINS
const ModuleFederationConfiguration = require('./module-federation');

// UTILS
const useWebpackCommonConfig = require('@goodt-webpack/common/webpack.common');
// CONST
const { RESOLVER_MODULE_NAME: REMOTE_RESOLVER_MODULE_NAME, REMOTE_ENTRY_FILENAME, APP_BUILD_ENTRY_PATH } = require('./const');

// prettier-ignore
const commonConfig = ({
  /* isEsm = true, */
  /* dependenciesMode = 'shared', // 'shared', 'external' */

  /* baseDir = process.env.INIT_CWD ?? process.env.PWD, */
  /* outputDir = './dist', */
  /* baseDir, */
  /* entry, */
  /* buildEntryFilePath = __dirname + '/build.js', */
  useShared = true,
  shareWcorePackages = true,

  projectDir,
  buildEntryFilePath = path.resolve(projectDir, APP_BUILD_ENTRY_PATH),

  packageJson = require(path.join(projectDir, 'package.json')),
  entry = packageJson.name,

  remoteConfig: {
    sharedConfig,
    entryFile,
    exposes
  } = {
    widgetName: REMOTE_RESOLVER_MODULE_NAME,
    entryFile: REMOTE_ENTRY_FILENAME,
    exposes: {
      [`./${REMOTE_RESOLVER_MODULE_NAME}`]: APP_BUILD_ENTRY_PATH
    }
  }
}) => {
  entry = entry.split('/').pop();

  return {
    entry: {
      [`${entry}`]: [ buildEntryFilePath ]
    },
    plugins: [
      ModuleFederationConfiguration({ packageJson, sharedConfig, exposes, entryFile, useShared, shareWcorePackages })
    ]
  };
};

module.exports = (config) => merge(useWebpackCommonConfig(config), commonConfig(config));
