// @goodt/scaffold/templates/basic/tpl/descriptor.js
/* eslint-disable no-restricted-syntax */
import panels from '[[{panelPath}]]';

/**
 * @enum {string}
 * @type {Readonly<Record<string, string>>}
 */
export const Vars = Object.freeze({
    [[#tips]]
    FOO: 'foo',
    BAR: 'bar'
    [[/tips]]
});

/**
 * @enum {string}
 * @type {Readonly<Record<string, string>>}
 */
export const Events = Object.freeze({
    [[#tips]]
    OPEN: 'open',
    CLOSE: 'close'
    [[/tips]]
});

/**
 * @description Don't change `descriptor` exported name
 * @return {ElemDescriptor}
 */
export const descriptor = () => ({
    props: {
        title: {
            type: String,
            default: 'Simple widget'
        },
        shouldWatchStore: {
            type: String,
            default: true
        },
    [[#hasTransport]]
        [[#tips]]
        apiBaseUrl: {
            type: String,
            default: 'https://reqres.in',
            label: 'Service api base url'
        },
        [[/tips]]
        orgStructureApiUrl: {
            type: String,
            default: 'https://goodt-dev.goodt.me:8480',
            label: 'Service OrgStructure Api Url'
        },
    [[/hasTransport]]
        events: {
            type: Object,
            default: () => ({
            [[#tips]]
                [Events.OPEN]: 'openEvent',
                [Events.CLOSE]: 'closeEvent'
            [[/tips]]
            })
        }
    },
    events: {
        [[#tips]]
        [Events.OPEN]: ['listen'],
        [Events.CLOSE]: ['listen', 'trigger']
        [[/tips]]
    },
    vars: Vars
});

export const meta = {
    descriptor,
    panels,
    isChildAllowed: false
    [[#tips]]
    /*
    cssVars,
    slotNames: ['default']
    */
    [[/tips]]
}

export default descriptor;
