// @goodt/postcss-config/postcss.config.js
const variables = require('postcss-advanced-variables');
const nested = require('postcss-nested');
const postcssBem = require('postcss-bem-fix');
const tailwindcss = require('tailwindcss');
const autoprefixer = require('autoprefixer');
/*
const px2rem = require('postcss-plugin-px2rem');
*/

const BASE_PLUGINS = [
    autoprefixer,
/*
    px2rem({
        rootValue: 16,
        propBlackList: ['composes'],
        selectorBlackList: [':root']
    })
*/
];

const POSTCSS_ONLY_PLUGINS = [
    variables,
    postcssBem({
        style: 'bem',
        separators: {
            modifier: '--'
        },
        shortcuts: {
            component: 'b',
            descendent: 'e',
            modifier: 'm'
        }
    }),
    nested,
    // tailwindcss
];

const postcssOnlyPluginsByEnv = (/* env */) => POSTCSS_ONLY_PLUGINS;
const basePluginsByEnv = (/* env */) => BASE_PLUGINS;

module.exports = (api) => {
    // `api.mode` - `mode` value of webpack, please read https://webpack.js.org/configuration/mode/
    // `api.webpackLoaderContext` - loader context for complex use cases
    // `api.env` - alias `api.mode` for compatibility with `postcss-cli`
    // `api.options` - the `postcssOptions` options    plugins
    const { file, env = 'development' } = api;

    // BASE PLUGINS BY ENV
    const basePlugins = basePluginsByEnv(env);
    const postcssPlugins = postcssOnlyPluginsByEnv(env);

    // VITE, DEVELOPMENT MODE
    const isVite = file == null;
    if (isVite) {
        return {
            plugins: [...basePlugins, ...postcssPlugins, tailwindcss]
        };
    }

    // WEBPACK
    const { options: { excludePatterns = [] } = {} } = api;

    const isOnlyBasePlugins = [
        // process css and less only with base
        /\.(less|css)$/.test(file),
        // or exclude files by pattern
        excludePatterns.some((re) => re.test(file))
    ].some(Boolean);

    // BASE PLUGINS
    if (isOnlyBasePlugins) {
        return {
            plugins: basePlugins
        };
    }

    const plugins = [...postcssPlugins, ...basePlugins];
    // use tailwindcss plugin once while processing tailwind.pcss to enable utility classes
    // all other tailwind features in other pcss files are no required
    if (file.includes('@goodt/postcss-config/tailwind.pcss')) {
        plugins.push(tailwindcss);
    }

    // BASE + POSTCSS-SPECIFIC PLUGINS
    return {
        plugins
    };
};
