// goodt-wcore/src/common/form/mixins/form.ts
export type FieldValidator = (val: string) => String | null;

export interface Form {
    context: FormContext;
    setValue(field: string, value: any);
    setErrors(field: string, errors: string[]);
    setValidators(field: string, validators: FieldValidator[]);
    submit();
    cancel();
    reset();
}

export interface FormContext {
    model: object;
    schema: { [field: string]: FormModelFieldSchema };
    validation: { [field: string]: FormModelFieldValidation };
    isLoading: boolean;
    isResetAvailable: boolean;
    isSubmitAvailable: boolean;
    hasErrors: boolean;
}

export interface FormModelFieldValidation {
    touched: boolean;
    errors: string[];
    validators: FieldValidator[];
}

export interface FormModelFieldOption {
    label: string;
    value: any;
}

export interface FormModelFieldSchema {
    type: string;
    isMultiple: Boolean;
    isReadOnly: boolean;
    isRequired: boolean;
    options: () => Promise<FormModelFieldOption[]>;
}

export interface StoreFilterItem {
    key: string;
    value: any;
}

export interface DescriptorFieldValue {
    value: string;
    type: number;
}
