// @goodt/scaffold/templates/basic/tpl/api/ApiService.js
import { ApiClientMethod, BaseApiService, processRequestResult, buildRequest } from '@goodt-common/api';

/**
 * Service actions metadata
 */
const ServiceAction = {
    GET_USER_BY_ID: {
        url: 'users/:id',
        // options is optional and could be omitted
        // default HTTP Request Method is ApiClientMethod.GET === 'GET' and also could be omitted
        options: { method: ApiClientMethod.GET }
    }
};

/**
 * ApiService class
 */
export class ApiService extends BaseApiService {
    /**
     *
     * @param {IApiServiceRequestOptions} request
     * @param [DtoConstructor]
     * @param [processOptions]
     * @return {Promise<SafeResult<*, Error>>}
     */
    async request(request, DtoConstructor, processOptions) {
        return processRequestResult(await super.request(buildRequest(request)), DtoConstructor, processOptions);
    }

    /**
     * Get user info by id
     *
     * @param {string|number} id user id
     * @return {Promise<import('@goodt-common/utils').ISafeResult<*, Error>>} result
     */
    getUserById(id) {
        // HTTP GET Method query string params
        const queryParams = {
            order: 'asc',
            sort: 'name'
        };

        /**
         * @example
         * for `id = 1`
         * final request `url` path with query string would be
         * `users/1?sort=name&order=asc`
         */

        return this.request({
            action: ServiceAction.GET_USER_BY_ID,
            pathParams: { id },
            params: queryParams
        });
    }
}
