// @goodt/webpack/host/config/module-federation.js
const AutomaticVendorFederation = require('@module-federation/automatic-vendor-federation');
const { getShared } = require('@goodt-webpack/common/shared');
const {
    container: { ModuleFederationPlugin }
} = require('webpack');

module.exports = ({ packageJson, packageLockJson, sharedConfig, useShared, isEsm, shareWcorePackages }) => {
    return new ModuleFederationPlugin({
        ...(isEsm && {
            library: {
                type: 'module'
            }
        }),
        // This should be our 'shared contract' between the host and the remotes.
        shared: {
            ...AutomaticVendorFederation({
                // exclude is mandatory prop
                exclude: [
                    ...packageJson.sharedDependencies,
                    '@goodt/common-deps',
                    '@goodt/editor-common',
                    '@goodt/package-loader'
                ],
                // ignoreVersion: [],
                packageJson,
                shareFrom: ['dependencies', 'peerDependencies']
            }),
            ...(useShared && getShared({ packageJson, packageLockJson, isHost: true, shareWcorePackages })),
            ...sharedConfig
        },
    });
};
