// @goodt/webpack/host/config/webpack.common.js
// LIBS
const path = require('path');
const { merge } = require('webpack-merge');

// PLUGINS
const ModuleFederationConfiguration = require('./module-federation');
const HtmlWebpackInjectPreload = require('@principalstudio/html-webpack-inject-preload');
const WebpackAssetsManifest = require('webpack-assets-manifest');
const HtmlWebpackInjectPlugin = require('html-webpack-inject-plugin').default;
const HtmlWebpackInjectAttributesPlugin = require('html-webpack-inject-attributes-plugin');
const CopyPlugin = require('copy-webpack-plugin');
// const HtmlWebpackPlugin = require('html-webpack-plugin');

// prettier-ignore
// UTILS
const useWebpackCommonConfig = require('@goodt-webpack/common/webpack.common');

// CONST
const { OUTPUT_DIR } = require('@goodt-webpack/common/const');
const { APP_BUILD_ENTRY_PATH, REMOTES_PATH } = require('./const');

const commonConfig = ({
  isEsm,
  isProd,
  useShared = true,
  shareWcorePackages = true,
  // useExternals = false,

  projectDir,
  baseDir = process.cwd(),
  outputDir = OUTPUT_DIR,
  buildEntryFilePath = path.resolve(projectDir, APP_BUILD_ENTRY_PATH),

  packageJson = require(path.join(projectDir, 'package.json')),
  entry = 'host',

  moduleFederation: {
    sharedConfig
  } = {}
}) => {

  // брать последнюю часть названия entry в случае наличия '/'
  entry = entry.split('/').pop();

  return {
    entry: {
      [`${entry}`]: [ buildEntryFilePath ]
    },
    plugins: [
      new CopyPlugin({
        patterns: [
          {
            from: path.resolve(baseDir, 'node_modules', '@goodt/editor-common/sw/workers'),
            to: path.resolve(baseDir, outputDir),
            toType: 'dir',
            info: { minimized: true }
          },
          {
            from: path.join(baseDir, 'node_modules', 'goodt-framework-css/dist/all.css'),
            to: path.resolve(baseDir, outputDir, 'css', 'goodt-framework-styles.css')
          },
          {
            from: path.join(baseDir, 'node_modules', '@mdi/font/css'),
            to: path.resolve(baseDir, outputDir, 'css'),
            toType: 'dir'
          },
          {
            from: path.join(baseDir, 'node_modules', '@mdi/font/fonts'),
            to: path.resolve(baseDir, outputDir, 'fonts'),
            toType: 'dir'
          },
          {
            from: path.join(baseDir, 'node_modules', '@mdi/light-font/css'),
            to: path.resolve(baseDir, outputDir, 'css'),
            toType: 'dir'
          },
          {
            from: path.join(baseDir, 'node_modules', '@mdi/light-font/fonts'),
            to: path.resolve(baseDir, outputDir, 'fonts'),
            toType: 'dir'
          },
          {
            from: path.join(baseDir, 'node_modules', '@goodt/webpack/host/public'),
            to: path.resolve(baseDir, outputDir, 'wmf'),
            toType: 'dir'
          }
        ]
      }),
      new HtmlWebpackInjectPreload({
        files: [
          {
            match: /\bgoodt-framework-styles.css$/,
            attributes: { rel: 'stylesheet', fetchPriority: 'high', as: 'style', /* onload: "this.rel = 'stylesheet'" */ },
          },
          {
            match: /\bmaterialdesignicons.min.css$/,
            attributes: { as: 'style', onload: "this.rel = 'stylesheet'" },
          },
          {
            match: /\bmaterialdesignicons-light.min.css$/,
            attributes: { as: 'style', onload: "this.rel = 'stylesheet'" },
          },
        ]
      }),
      new HtmlWebpackInjectPlugin({
        externals: [
          {
            tagName: 'script',
            attributes: {
              src: 'wmf/wmf.utils.js',
              type: 'module',
              defer: 'defer'
            }
          },
          // },
          // @todo keep for reference
          // @todo not used since @goodt/webpack@1.7.0
          // {
          //   tagName: 'style',
          //   innerHTML: '@import url("css/goodt-framework-styles.css")',
          //   type: null
          //
          // {
          //   tagName: 'script',
          //   attributes: {
          //     src: 'wmf/entries.bootstrap.js',
          //     type: 'module',
          //     defer: 'defer'
          //   }
          // },
          // {
          //   tagName: 'script',
          //   attributes: {
          //     src: 'wmf/wmf.bootstrap.js',
          //     defer: 'defer'
          //   }
          // },
          // ...(useExternals
          //     ? [
          //       {
          //         tagName: 'script',
          //         attributes: {
          //           preload: false,
          //           src: 'wmf/importmap.json',
          //           type: 'importmap'
          //         }
          //       },
          //       // 2nd to support esm-shim
          //       {
          //         tagName: 'script',
          //         attributes: {
          //           preload: false,
          //           async: true,
          //           src: 'https://unpkg.com/es-module-shims'
          //         }
          //       }
          //     ] : []
          // ),
        ],
        prepend: true // default is false
      }),
      new HtmlWebpackInjectAttributesPlugin({
        type: 'module'
      }),  //
      new WebpackAssetsManifest({}),
      ModuleFederationConfiguration({
        isEsm,
        isProd,
        packageJson,
        sharedConfig,
        useShared,
        shareWcorePackages
      })
    ]
  };
};

module.exports = (config, webpackConfig) => merge(useWebpackCommonConfig(config, webpackConfig), commonConfig(config, webpackConfig));
