// @goodt/webpack/common/optimization.js
const getPackageName = (module) => {
    const regExp = /\/node_modules\/(.*?)(\/|$)([^\/]*)?/;
    const { context, _identifier } = module;

    for (const str of [context, _identifier]) {
        const strMatch = str.match(regExp);
        if (strMatch != null && strMatch.length > 1) {
            if (strMatch[1].startsWith('@') || !['', 'src', undefined].includes(strMatch[3])) {
                strMatch[1] += '_' + strMatch[3];
            }
            return strMatch[1];
        }
    }

    return 'unnamed';
};

const createOptimization = (/*{ splitChunks: { cacheGroups = {} } = {}} = {}*/) => ({
    runtimeChunk: false,
    splitChunks: {
        cacheGroups: {
            // ...cacheGroups,
            vendors: {
                chunks: 'async',
                test: /\/node_modules\//,
                reuseExistingChunk: true,
                maxSize: 300 * 1000,
                minChunks: 2,
                /*
                  The higher priority will determine where a module is placed
                  if it meets multiple conditions (both a shared and npm (vendor) module

                  Prioritize vendor chunks over commons
                */
                priority: 20,
                name(module) {
                    const packageName = getPackageName(module);
                    return `npm.${packageName.replace('@', '')}`;
                },
                //     /*
                //       Enforce value is set to true to force SplitChunksPlugin to
                //       form this chunk irrespective of the size of the chunk
                //     */
                enforce: true
            },
            // common: {
            //     // create a commons chunk, which includes all code shared between entry points
            //     name: `common`,
            //     // minimum number of chunks that must share a module before splitting
            //     minChunks: 2,
            //     // async + async chunks
            //     chunks: 'async',
            //     // lower priority than vendors
            //     priority: 10,
            //     /*
            //       If the current chunk contains modules already split out from the main bundle,
            //       it will be reused instead of a new one being generated.
            //     */
            //     reuseExistingChunk: true,
            //     /*
            //       Enforce value is set to true to force SplitChunksPlugin to
            //       form this chunk irrespective of the size of the chunk
            //     */
            //     enforce: true
            // }
        }
    }
});

module.exports = () => {
    return createOptimization();
};
