// @goodt/webpack/common/webpack.common.js
// LIBS
const { pick } = require('lodash');
const { argv } = require('./utils');
const { resolve } = require('path');

// PLUGINS
const { DefinePlugin } = require('webpack');
// const DynamicContainerPathPlugin = require('dynamic-container-path-webpack-plugin');
const CircularDependencyPlugin = require('circular-dependency-plugin');

// CONFIGS
const Rules = require('./rules');
const Externals = require('./externals');

// CONSTS
const { ARGV_DEV_TOOLS, OUTPUT_DIR } = require('./const');
const { REMOTES_PATH } = require('../host/config/const');
const { REMOTE_ENTRY_FILENAME } = require('../remote/config/const');
// const Aliases = require('./aliases');

// prettier-ignore
const commonConfig = ({
  isEsm = true,
  isProd,

  useExternals = false,
  useDevTools = false,

  projectDir,
  outputDir = OUTPUT_DIR,

  packageJson = require(resolve(projectDir, 'package.json')),
  entry = packageJson.name,
}) => {
    // брать последнюю часть названия entry в случае наличия '/'
    entry = entry.split('/').pop();

    return {
        experiments: {
            ...isEsm && { outputModule: true },
            topLevelAwait: true
        },
        target: 'web',
        output: {
            publicPath: 'auto',
            path: outputDir,
            clean: true,
            ...(isProd
                ? {
                    // will be enabled by defaults by experiments.outputModule = true
                    // module: true,
                    // chunkFormat: 'module',
                    // enabledChunkLoadingTypes: [ 'import', 'jsonp' ]
                    chunkFilename: `js/${entry}.[name].[chunkhash:6].js`,
                }
                : {
                    // fix HMR in 'development'
                    chunkFormat: 'array-push'
                })
        },
        module: {
            rules: [ ...Rules ]
        },
        resolve: {
            alias: {
                lodash: 'lodash-es'
                //...Aliases
            },
            fallback: {
                'querystring': require.resolve('querystring-es3'),
                'path': require.resolve('path-browserify')
            }
        },
        optimization: {
            /*
              disable webpack base config `runtimeChunk: single`
              https://github.com/webpack/webpack/issues/11691

              This can be removed when #11843 is merged
              https://github.com/webpack/webpack/pull/11843
            */
            runtimeChunk: false
        },
        // when use externals
        ...(useExternals && {
            externalsType: 'module',
            externals: pick({
                ...Externals.VENDORS,
                ...Externals.CORE
            }, packageJson.externalDependencies)
        }),
        plugins: [
            new DefinePlugin({
                // Drop Options API from bundle
                __VUE_PROD_DEVTOOLS__: JSON.stringify(isProd !== true || useDevTools || argv(ARGV_DEV_TOOLS) != null),
                '__WMF__.USE_ESM': JSON.stringify(isEsm),
                '__WMF__.REMOTES_PATH': JSON.stringify(REMOTES_PATH),
                '__WMF__.REMOTE_ENTRY_FILENAME': JSON.stringify(REMOTE_ENTRY_FILENAME)
            }),
            new CircularDependencyPlugin({
                // exclude detection of files based on a RegExp
                exclude: /node_modules/,
                // include specific files based on a RegExp
                // add errors to webpack instead of warnings
                failOnError: true,
                // allow import cycles that include an asynchronous import,
                // e.g. via import(/* webpackMode: "weak" */ './file.js')
                allowAsyncCycles: true,
                // set the current working directory for displaying module paths
                cwd: process.cwd()
            })
            /*
            // used when webpackConfig.output.publicPath not equal 'auto'
            new DynamicContainerPathPlugin({
                // provide the code to get `publicPath` at runtime
                iife: function (entry) {
                    const { ENTRIES_MAP, ENV } = window.__GOODT__.WMF;
                    const { url } = ENTRIES_MAP[entry][ENV];
                    const publicPath = url + "/";
                    return publicPath;
                },
                /!*
                  Provide the main entry point as an argument to the plugin above. The value will be
                  provided as a key to `map.config.json` to get the URL for this app
                *!/
                entry
            })
            */
        ]
    };
};
module.exports = commonConfig;
