// goodt-dremio-sdk/src/constants.js
/**
 * @namespace
 */
export const Dremio = {
    /**
     * @enum {String}
     */
    DATASET_TYPE: {
        /** @member {String} */
        VIRTUAL: 'VIRTUAL',
        /** @member {String} */
        PROMOTED: 'PROMOTED',
        /** @member {String} */
        DIRECT: 'DIRECT'
    },
    /**
     * @enum {String}
     */
    CONTAINER_TYPE: {
        /** @member {String} */
        SPACE: 'SPACE',
        /** @member {String} */
        SOURCE: 'SOURCE',
        /** @member {String} */
        FOLDER: 'FOLDER',
        /** @member {String} */
        HOME: 'HOME'
    },
    /**
     * @enum {String}
     */
    JOB_STATUS: {
        /** @member {String} */
        NOT_SUBMITTED: 'NOT_SUBMITTED',
        /** @member {String} */
        STARTING: 'STARTING',
        /** @member {String} */
        RUNNING: 'RUNNING',
        /** @member {String} */
        COMPLETED: 'COMPLETED',
        /** @member {String} */
        CANCELED: 'CANCELED',
        /** @member {String} */
        FAILED: 'FAILED',
        /** @member {String} */
        CANCELLATION_REQUESTED: 'CANCELLATION_REQUESTED',
        /** @member {String} */
        ENQUEUED: 'ENQUEUED'
    },
    /**
     * @enum {String}
     */
    SOURCE_TYPE: {
        /** @member {String} */
        AMAZONELASTIC: 'Amazon Elasticsearch Service',
        /** @member {String} */
        REDSHIFT: 'Amazon Redshift',
        /** @member {String} */
        S3: 'Amazon S3',
        /** @member {String} */
        ADLS: 'Azure Data Lake Storage Gen1',
        /** @member {String} */
        AZURE_STORAGE: 'Azure Storage',
        /** @member {String} */
        ELASTIC: 'Elasticsearch',
        /** @member {String} */
        HDFS: 'HDFS',
        /** @member {String} */
        HIVE: 'Hive',
        /** @member {String} */
        MAPRFS: 'MapR-FS',
        /** @member {String} */
        MSSQL: 'Microsoft SQL Server',
        /** @member {String} */
        MONGO: 'MongoDB',
        /** @member {String} */
        MYSQL: 'MySQL',
        /** @member {String} */
        NAS: 'NAS',
        /** @member {String} */
        ORACLE: 'Oracle',
        /** @member {String} */
        POSTGRES: 'PostgreSQL'
    }
};
