// @goodt/webpack/host/chainWebpack.js
const path = require('path');
module.exports.build = ({
    entry,
    projectDir,
    useStyleShadowMode = false
    /*
    baseDir,
    outputDir: OUTPUT_DIR,
    buildEntryFilePath: APP_BUILD_ENTRY_PATH,
    packageJson
    */
}) => (chainableConfig) => {
    // под вопросом – необходимо для goodt-edition-builder,
    // потому что собираем через VUE_CLI_CONTEXT из рута проекта goodt-edition-builder
    // ./none_modules/@goodt/editor, ./none_modules/@goodt/player
    chainableConfig.context(projectDir);

    /*
    if (useStyleShadowMode) {
        const enableShadowMode = require('@goodt-webpack/common/rules/shadow-mode');
        enableShadowMode(chainableConfig);
    }
    */

    chainableConfig.plugins.delete('prefetch');
    chainableConfig.plugins.delete('preload');
    // @see https://cli.vuejs.org/guide/html-and-static-assets.html#preload
    chainableConfig.plugins.delete('case-sensitive-paths');

    // под вопросом – необходимо для goodt-edition-builder,
    // потому что собираем через VUE_CLI_CONTEXT из рута проекта goodt-edition-builder
    // ./none_modules/@goodt/editor, ./none_modules/@goodt/player
    chainableConfig.module
      .rule('js')
      .use('babel-loader')
      .loader('babel-loader')
      .options({
          presets: [
              [require.resolve('@vue/babel-preset-app')]
          ]
      });

    chainableConfig.optimization
      .minimizer('terser')
      .tap(([ options, ...rest ]) => [ {
          ...options,
          terserOptions: {
              ...options.terserOptions,
              sourceMap: {
                  file: '[name].map'
              },
              // process modules with top-level await support
              module: true
          }
      }, ...rest ]);

    // WMF-specific webpack config manipulations
    chainableConfig.optimization
      // possibly delete Terser to replace with EsbuildPlugin in `webpack.prod.js`
      // .minimizers.delete('terser')
      // remove vue-cli minimize settings
      // .minimize(false)
      .delete('splitChunks');

    // @see https://cli.vuejs.org/guide/html-and-static-assets.html#preload
    // exclude "main" entry point from be injected in html
    chainableConfig.entryPoints.store.delete('app');

    if (entry != null) {
        // exclude "main" chunks from be injected in html
        // @todo вероятно нам это ненужно
        chainableConfig.plugin('html').init((Plugin, args) => {
            args[0] = {
                ...args[0],
                excludeChunks: [entry]
            };

            return new Plugin(...args);
        });
    }
};
