// goodteditor-ui/styleguide.config.js
const path = require('path');
const { version } = require('./package.json');

module.exports = {
    components: 'src/components/ui/**/[A-Z]*.vue',
    require: ['goodt-framework-css'],
    title: 'UI',
    version,
    styleguideDir: 'docs',
    getComponentPathLine(componentPath) {
        const name = path.basename(componentPath, '.vue');
        return `import { ${name} } from 'goodteditor-ui';`;
    },
    ignore: [
        'src/components/ui/ColorPicker/**',
        'src/components/ui/utils/**',
        'src/components/ui/WysiwygEditor/extensions/**',
        'src/components/ui/WysiwygEditor/renders/**',
    ],
    template: {
        head: {
            links: [
                {
                    rel: 'stylesheet',
                    href:
                        'https://cdn.materialdesignicons.com/5.4.55/css/materialdesignicons.min.css',
                    crossorigin: 'anonymous',
                },
                {
                    rel: 'stylesheet',
                    href: 'https://fonts.googleapis.com/css2?family=Roboto:wght@400;500&display=swap',
                    crossorigin: 'anonymous',
                },
                {
                    rel: 'stylesheet',
                    href:
                        'https://materialdesignicons.com/cdn/light/0.2.63/css/materialdesignicons-light.min.css',
                    crossorigin: 'anonymous',
                },
            ],
        },
        body: {
            raw: `<script src="https://unpkg.com/dayjs@1.9.6/dayjs.min.js"></script>
<script src="https://unpkg.com/dayjs@1.9.6/locale/ru.js"></script>
<script src="https://unpkg.com/dayjs@1.9.6/plugin/customParseFormat.js"></script>`,
        },
    },
};
