// @goodt/scaffold/templates/dataset/tpl/descriptor.js
[[#customPanel]]
import { useDatasetMeta } from '@goodt-common/data';
import { DatasetPanelMixin } from '[[{panelPath}]]';
[[/customPanel]]

import panels from '[[{panelPath}]]';

[[#additionalDataset]]
export const DATASET_ADDITIONAL_PROP = 'additionalDatasets';
[[/additionalDataset]]

/**
 * @description Don't change `descriptor` exported name
 * @return {ElemDescriptor}
 */
export const descriptor = () => ({
    props: {
        metric: {
            type: String,
            default: null,
            label: 'Метрика'
        },
        dimension: {
            type: String,
            default: null,
            label: 'Измерение'
        },
        dimensionOrMetric: {
            type: String,
            default: null,
            label: 'Измерение/метрика'
        },
        field: {
            type: String,
            default: null,
            label: 'Поле'
        },
        [[#multiple]]
        mainDimensionOrMetric: {
            type: String,
            default: null,
            label: 'Главное измерение/метрика'
        },
        mainDatasetIndex: {
            type: Number,
            default: 0,
            label: 'Главный датасет'
        },
        [[/multiple]]
        [[#drilldown]]
        drilldownName: {
            type: String,
            default: null,
            label: 'Дриллдаун'
        },
        [[/drilldown]]
    }
});

[[#simpleMeta]]
export const meta = {
[[/simpleMeta]]
[[^simpleMeta]]
export const metaBase = {
[[/simpleMeta]]
    descriptor,
    panels,
    isChildAllowed: false
    /*
    cssVars,
    slotNames: ['default']
    */
};

[[#customPanel]]
export const meta[[#additionalDataset]]BaseWithCustomPanel[[/additionalDataset]] = useDatasetMeta(metaBase, {
    panel: {
        // meta: { name: '', icon: '' },
        // isMultiple: false // (false by default)
        // DatasetPanelMixin расширяет поведение стандартной DatasetPanel.vue
        mixins: [DatasetPanelMixin]
    }
});
[[/customPanel]]

[[#additionalDataset]]
export const meta = useDatasetMeta(metaBase[[#customPanel]]WithCustomPanel[[/customPanel]], {
    prop: DATASET_ADDITIONAL_PROP,
    panel: {
        meta: { name: 'Дополнительный источник' }
        // isMultiple: false // (false by default)
    }
});
[[/additionalDataset]]
