// @goodt/webpack/remote/chainWebpack.js
// prettier-ignore
const enableShadowMode = require('@goodt-webpack/common/rules/shadow-mode');

module.exports.build = ({
    /*
    baseDir,
    outputDir: OUTPUT_DIR,
    buildEntryFilePath: APP_BUILD_ENTRY_PATH,
    packageJson,
    remoteConfig: {
      name: containerName,
      entryFile: REMOTE_ENTRY_NAME,
      exposes: {
      }
    }
    */
}) => (chainableConfig) => {
    chainableConfig.entryPoints.store.delete('app');

    chainableConfig
      .module
      .rules
      .delete('js');

    chainableConfig.optimization
      // possibly delete Terser to replace with EsbuildPlugin in `webpack.prod.js`
      // .minimizers.delete('terser')
      .delete('splitChunks');

    chainableConfig.plugins.delete('html');

    /*
    chainableConfig.plugin('html').init((Plugin, args) => {
        args[0] = {
            ...args[0],
            excludeChunks: [...entriesConfig.entrypoints]
        };

        return new Plugin(...args);
    });
    */

    // enable shadow mode
    // used for editor in-realm (iframe) preview mode
    // @todo temporary disabled
    // also required applying enableShadowMode(chainableConfig) in host webpack config
    // enableShadowMode(chainableConfig);
};
