// @goodt/scaffold/templates/dataset/index.js
const Template = require('../template.js');
const { Select } = require('enquirer');

module.exports = class extends Template {
    constructor(widgetName, config) {
        super(widgetName, config);
    }

    get path() {
        return __dirname;
    }

    async build(options) {
        super.build(options);

        const tplType = await new Select({
            message: 'Single/Multiple Data Sources',
            choices: Object.entries({
                single: 'single',
                multiple: 'multiple'
            }).map(([name, value]) => ({
                value,
                name
            }))
        }).run();

        const drilldown =
            (await new Select({
                message: 'Use drilldown',
                choices: Object.entries({
                    yes: 'yes',
                    no: 'no'
                }).map(([name, value]) => ({
                    name,
                    value
                }))
            }).run()) === 'yes';

        const customPanel =
            (await new Select({
                message: 'Use Dataset Panel Mixin',
                choices: Object.entries({
                    yes: 'yes',
                    no: 'no'
                }).map(([name, value]) => ({
                    value,
                    name
                }))
            }).run()) === 'yes';

        let additionalDataset = false;
        if (tplType === 'simple') {
            additionalDataset =
                (await new Select({
                    message: 'Use Additional Dataset Panel',
                    choices: Object.entries({
                        yes: 'yes',
                        no: 'no'
                    }).map(([message, value]) => ({
                        value,
                        message
                    }))
                }).run()) === 'yes';
        }

        const tips =
            (await new Select({
                message: 'Use "Tips"',
                choices: Object.entries({
                    yes: 'yes',
                    no: 'no'
                }).map(([message, value]) => ({
                    value,
                    message
                }))
            }).run()) === 'yes';

        const tplBindings = {
            core: this.corePath,
            corePanels: this.corePanelsPath,
            coreMixins: this.coreMixinsPath,
            lib: this.widgetLibPath,
            path: this.widgetPath,
            name: this.widgetName,
            panelName: this.config.panel.name,
            panelPath: this.config.panel.path,

            multiple: tplType === 'multiple',
            tips,
            drilldown,
            customPanel,
            additionalDataset,
            simpleMeta: !customPanel && !additionalDataset
        };

        return this.createWidget({
            tplBindings,
            tplType: { elem: tplType }
        });
    }

    /**
     * Create widget files (elem, panel, dirs)
     * @param {object} tpl  compiled vue template files
     * @return {boolean|Error}  true if success; else error
     */
    createWidget({ tplBindings, tplPath = this.tplPath, tplType }) {
        super.createWidget({ tplBindings, tplPath, tplType });

        const { path: panelPath } = this.config.panel;
        if (tplBindings.customPanel) {
            this.buildWidgetFile(`${tplPath}/DatasetPanelMixin.js`, `${panelPath}/DatasetPanelMixin.js`, tplBindings);
        }

        return true;
    }
};
