// @goodt/sandbox/vite.config.js
const { join } = require('path');
const { createVuePlugin } = require('vite-plugin-vue2');
const { defineConfig } = require('vite');
const { include: OptimizeDepsInclude } = require('./optimizeDeps');
const WCorePackageAliases = require('@goodt-wcore/root/package.aliases');
const { getConfigFiles, publicDir, BASE_DIR } = require('./utils');
const aliases = require('./aliases');

// eslint-disable-next-line import/no-dynamic-require
const {
    dependencies = {},
    // eslint-disable-next-line unicorn/prevent-abbreviations
    devDependencies = {},
    peerDependencies = {},
    optionalDependencies = {},
    viteOptimizeExcludeDependencies = []
} = require(process.env.npm_package_json);

// eslint-disable-next-line no-restricted-syntax
const ExcludedDepsByPackageJson = [dependencies, devDependencies, peerDependencies, optionalDependencies]
    .flatMap((deps) => Object.keys(deps))
    .filter((dep) => viteOptimizeExcludeDependencies.some((pattern) => new RegExp(pattern).test(dep)));

/**
 * @type {import('vite').UserConfig}
 */
module.exports = defineConfig({
    plugins: [createVuePlugin()],
    publicDir,
    root: __dirname,
    server: {
        cors: {
            origin: '*'
        },
        fs: {
            allow: [BASE_DIR]
        }
    },
    define: {
        process: {
            env: {
                ...process.env,
                WIDGET_FILES: getConfigFiles(publicDir)
            }
        }
    },
    optimizeDeps: {
        include: OptimizeDepsInclude,
        exclude: [
            'goodt-wcore',
            'vue',
            'goodteditor-ui',
            ...ExcludedDepsByPackageJson,
            ...Object.keys(WCorePackageAliases)
        ],
        // https://github.com/SortableJS/vue.draggable.next/issues/49
        // https://github.com/SortableJS/Vue.Draggable/issues/738
        esbuildOptions: { keepNames: true }
    },
    resolve: {
        alias: [
            // { find: /(.+)\?v=(.*)/gim, replacement: '$1' },
            { find: /^(@root)\/(.*)/, replacement: join(BASE_DIR, '$2') },
            { find: /^vuedraggable/, replacement: 'vuedraggable/src/vuedraggable' },
            ...aliases
        ]
    },
    css: {}
});
