// @goodt/sandbox/webpack.sandbox.js
const { DefinePlugin } = require('webpack');
const CopyPlugin = require('copy-webpack-plugin');
const { getConfigFiles, publicDir } = require("./utils");
const Rules = require('./rules');

const BASE_DIR = process.env.PWD ?? process.env.INIT_CWD;

module.exports.configureWebpack = (config) => {
    config.module.rules.push(...Rules);

    config.plugins.push(
        new CopyPlugin({
            patterns: [
                {
                    from: publicDir,
                    to: '.',
                    toType: 'dir'
                }
            ]
        }),
        new DefinePlugin({
            'process.env.WIDGET_FILES': JSON.stringify(getConfigFiles(publicDir)),
        })
    );

    config.entry = {
        app: [require.resolve('@goodt/sandbox')]
    };

    config.resolve.alias = {
        ...config.resolve.alias,
        '@root': BASE_DIR
    };

    config.resolve.fallback = {
        ...config.resolve.fallback,
        path: require.resolve('path-browserify')
    };
};
