// @goodt/posthtml-config/index.js
const { default: bemify } = require('posthtml-bemify');
const svgTags = require('svg-tags');
const { nodeVisitor } = require('./utils');

// Tags to be skipped in processing with children
const SkipTags = svgTags;
// Tags to be ignored in processing if no extra processing attributes specified
const IgnoreTags = ['select', 'div', 'span', 'slot', 'a', 'p', 'svg', 'template', 'transition'];
// Specific tags to concrete class
const ClassesMap = {
    h1: 'title',
    ul: 'list',
    li: 'list-item',
    a: 'link',
    p: 'text'
};

const CommonOptions = {
    skipTags: SkipTags,
    ignoreTags: IgnoreTags,
    // Tags to skip transform tag in preprocess to custom
    // cause it Components name
    ignoreTransformTag: { tag: /^(w|ui)-/ },
    nodeVisitor
};
// prettier-ignore

module.exports = {
    bem: {
        plugins: [bemify({
            ...CommonOptions,
            // specific attributes matching on BEM parts
            matcher: {
                BLOCK: '~b',
                ELEMENT: '~e',
                MODIFIER: '~m',
                ELEMENT_OF: "~e:of",
                SKIP_CHILDREN: "~~"
            }
        })]
    },
    tag: {
        plugins: [bemify({
            ...CommonOptions,
            bem: false,
            matcher: {
                BLOCK: { tag: /^[a-z-]{3,}$/ },
                MODIFIER: '~m'
            }
        })]
    }
};
